homeassistant/components/switch/mochad.py

```diff
@@ -0,0 +1,81 @@
+"""
+Contains functionality to use a X10 switch over Mochad.
+
+For more details about this platform, please refer to the documentation at
+https://home.assistant.io/components/switch.mochad
+"""
+
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components import mochad
+from homeassistant.components.switch import SwitchDevice
+from homeassistant.const import (CONF_NAME, CONF_PLATFORM)
+from homeassistant.helpers import config_validation as cv
+
+DEPENDENCIES = ['mochad']
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ADDRESS = 'address'
+CONF_DEVICES = 'devices'
+
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_PLATFORM): mochad.DOMAIN,
+    CONF_DEVICES: [{
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Required(CONF_ADDRESS): cv.x10_address,
+        vol.Optional('comm_type'): cv.string,
+    }]
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup X10 switches over a mochad controller."""
+    devs = config.get(CONF_DEVICES)
+    add_devices([MochadSwitch(
+        hass, mochad.CONTROLLER.ctrl, dev) for dev in devs])
+    return True
+
+
+class MochadSwitch(SwitchDevice):
+    """Representation of a X10 switch over Mochad."""
+
+    def __init__(self, hass, ctrl, dev):
+        """Initialize a Mochad Switch Device."""
+        from pymochad import device
+
+        self._controller = ctrl
+        self._address = dev[CONF_ADDRESS]
+        self._name = dev.get(CONF_NAME, 'x10_switch_dev_%s' % self._address)
+        self._comm_type = dev.get('comm_type', 'pl')
+        self.device = device.Device(ctrl, self._address,
+                                    comm_type=self._comm_type)
+        self._state = self._get_device_status()
```

Comment:
So I was planning to do that originally, but when I did that I was having a lot of trouble getting it to stay in the right state after flipping the switch. The state would jump around a lot and it wasn't behaving consistently. I think this is probably related to being a single socket interface mochad is serializing the requests and parallel requests aren't being handled so well. So when doing a lot of device switching around the same time, things had to wait for a periodic state update to actually get the right state. Just letting home-assistant handle the state was much more reliable in my testing.

Also, none of the x10 switches I have here are 2 way so whatever state is being returned by mochad is just from whatever command mochad last received. (I think mochad would query the device for its state if a 2 way device is used, but I don't have any to test that)